src/components/actions/processTrigger.ts

```diff
@@ -0,0 +1,101 @@
+
+  // Get Actions from matching Template (and match templateActionCode if applicable)
+  const actions = (await DBConnect.getActionsByTemplateId(templateId, trigger))
+    .filter((action) => {
+      if (!event_code) return true
+      else return action.event_code === event_code
+    })
+    .map((action) => (action.code !== 'alias' ? action : swapOutAliasedActions(templateId, action)))
+
+  const resolvedActions = await Promise.all(actions)
+
+  // Separate into Sequential and Async actions
+  const actionsSequential: ActionSequential[] = []
+  const actionsAsync: ActionInTemplate[] = []
+  for (const action of resolvedActions) {
+    if (action.sequence) actionsSequential.push(action as ActionSequential)
+    else actionsAsync.push(action)
+  }
+
+  for (const action of [...actionsAsync, ...actionsSequential]) {
+    // Add all actions to Action Queue
+    await DBConnect.addActionQueue({
+      trigger_event: trigger_id,
+      trigger_payload: payload,
+      template_id: templateId,
+      sequence: action.sequence,
+      action_code: action.code,
+      parameter_queries: action.parameter_queries,
+      parameters_evaluated: {},
+      condition_expression: action.condition,
+      status:
+        typeof action.sequence === 'number'
+          ? ActionQueueStatus.Processing
+          : ActionQueueStatus.Queued,
+    })
+  }
+  await DBConnect.updateTriggerQueueStatus({
+    status: TriggerQueueStatus.ActionsDispatched,
+    id: trigger_id,
+  })
+
+  // Get sequential Actions from database
+  const actionsToExecute = await DBConnect.getActionsProcessing(templateId)
+
+  // Collect output properties of actions in sequence
+  // "data" is stored output from scheduled triggers or verifications
+  let outputCumulative = { ...data?.outputCumulative }
+
+  // Execute sequential Actions one by one
+  let actionFailed = ''
+  for (const action of actionsToExecute) {
+    if (actionFailed) {
+      await DBConnect.executedActionStatusUpdate({
+        status: ActionQueueStatus.Fail,
+        error_log: 'Action cancelled due to failure of previous sequential action ' + actionFailed,
+        parameters_evaluated: null,
+        output: null,
+        id: action.id,
+      })
+      continue
+    }
+    try {
+      const actionPayload = {
+        id: action.id,
+        code: action.action_code,
+        condition_expression: action.condition_expression as EvaluatorNode,
+        parameter_queries: action.parameter_queries,
+        trigger_payload: action.trigger_payload,
+      }
+      const result = await executeAction(actionPayload, actionLibrary, {
+        outputCumulative,
+      })
+      outputCumulative = { ...outputCumulative, ...result.output }
+      // Debug helper console.log to inspect action outputs:
+      if (showActionOutcomeLog) console.log('outputCumulative:', outputCumulative)
+      if (result.status === ActionQueueStatus.Fail) console.log(result.error_log)
```

Comment:
should we be setting actionFailed = action.action_code here ?

Reply:
Hmm.. yes, strictly we probably should. However I'm slightly reluctant to change it now as I don't want to inadvertently break an existing template. 

From what I remember from debugging the action queue, most of the time an error is thrown, so it takes the "catch" route and writes actionFailed to true (therefore preventing subsequent actions from running). I'm basing that on the observation that I pretty much always see `Action cancelled due to failure of previous sequential action` in the error log of the action_queue table after a failed action.

However, I'll keep an eye on this as I do the next lot of configs and see if I'm correct -- and if I'm wrong whether there'd be any negative consequences. Then I'll come back and change this once I'm sure it's all good.

Reply:
Ok, i take your point about existing config might fail, but silently ignored error might be worse, as the side effects are only observed way further up the track (and at this stage it's hard to find the root).